### PR TITLE
feat: migrate current harness tests to sqlness

### LINKED ISSUE #397: Migrate current harness tests to sqlness
### Describe This Problem

[sqlness](https://github.com/CeresDB/sqlness) has been created as a separated repo, but ceresdb doesn't use it now.

### Proposal

Let's migrate the harness tests to the sqlness.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dust1**
# Which issue does this PR close?

Closes #397 

# Rationale for this change
 
migrate the harness tests to the sqlness.
* add a new dependency: `sqlness = { git = "https://github.com/CeresDB/sqlness.git", rev = "c077b17d73ab25460c152dc34e8f80f904522a57"}`
* update tokio version '1.0' to '1.21'

# What changes are included in this PR?

the tests dir. I create a new module use `sqlness`.

# Are there any user-facing changes?

none

# How does this change test

### CODE REVIEW

**Comment by ShiKaiWi** [comment]
@dust1 Would you like to do a rebase on the main branch to resolve the conflicts?

File: tests/sqlness/src/setup.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            )
+        });
+
+        let client = Builder::new(endpoint, Mode::Standalone).build();
+        Client::new(client)
+    }
+
+    /// Stop one [`Database`].
+    async fn stop(&self, _mode: &str, _database: Self::DB) {
+        // self.server_process.kill().unwrap()
```
Comment (ShiKaiWi) [inline tests/sqlness/src/setup.rs:41]: Maybe you have to rebase the main branch. After all, the #461 have fixed one bug for test failure.

**Review by ShiKaiWi — COMMENTED** [review]
I guess the contents of `harness` package should be replaced by contents of the `sqlness` package. Is that feasible?

File: tests/sqlness/src/setup.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+use crate::client::Client;
+
+const BINARY_PATH_ENV: &str = "CERESDB_BINARY_PATH";
+const CONFIG_PATH_ENV: &str = "CERESDB_CONFIG_PATH";
+const SERVER_ENDPOINT_ENV: &str = "CERESDB_SERVER_ENDPOINT";
+const CERESDB_STDOUT_FILE: &str = "CERESDB_STDOUT_FILE";
+const CERESDB_STDERR_FILE: &str = "CERESDB_STDERR_FILE";
+
+pub struct Environment {
```
Comment (jiacai2050) [inline tests/sqlness/src/setup.rs:19]: How about name this CeresDBEnv
```
impl Environment for CeresDBEnv {
  ......
}
```

File: tests/sqlness/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+[package.edition]
+workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+anyhow = "1.0.58"
+async-trait = "0.1"
+ceresdb-client-rs = { git = "https://github.com/CeresDB/ceresdb-client-rs.git", rev = "a9d9190f4f7b55171ea2ad142fb41dc9909c19c5" }
+prettydiff = { version = "0.6.1", features = [ "prettytable-rs"]}
```
Comment (jiacai2050) [inline tests/sqlness/Cargo.toml:14]: This dependency is not necessary.

File: tests/sqlness/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+sqlness = { git = "https://github.com/CeresDB/sqlness.git", rev = "c077b17d73ab25460c152dc34e8f80f904522a57"}
+tokio = { workspace = true }
+walkdir = "2.3.2"
```
Comment (jiacai2050) [inline tests/sqlness/Cargo.toml:17]: Same as above

**Comment by ShiKaiWi** [comment]
@dust1 I find that the harness test in the CI is broken without any information. Maybe you can run the harness test locally to check what is wrong.

**Comment by dust1** [comment]
🤔the ci is check lock file fail.
```
< name = "ceresdb-test"
---
> name = "ceresdb-test-sqlness"
```
i renamed the sqlness build name to `ceresdb-test`, but it failed

**Comment by ShiKaiWi** [comment]
> the ci is check lock file fail.
> 
> ```
> < name = "ceresdb-test"
> ---
> > name = "ceresdb-test-sqlness"
> ```
> 
> i renamed the sqlness build name to `ceresdb-test`, but it failed;😥

This failure is caused by the conflicts with the main branch. Maybe you should rebase the main branch and resolve this conflict firstly.

**Comment by dust1** [comment]
> 

ok

**Comment by dust1** [comment]
the ci was fail. I found the log is:
```
System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-111225-utc.log'
```
in my local, I use `cd tests && make run` was success.😥

**Comment by ShiKaiWi** [comment]
> the ci was fail. I found the log is:
> 
> ```
> System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-111225-utc.log'
> ```
> 
> in my local, I use `cd tests && make run` was success.😥

Oh, it seems the disk of the github workflow runner was exhausted.

**Comment by ShiKaiWi** [comment]
> > > the ci was fail. I found the log is:
> > > ```
> > > System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-111225-utc.log'
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > in my local, I use `cd tests && make run` was success.😥
> > 
> > 
> > Oh, it seems the disk of the github workflow runner was exhausted. Let me rerun the ci for a try.
> 
> It seems to have failed again😂

I can't see why it failed, and I'll give another try with enabling logging.

**Comment by dust1** [comment]
> I can't see why it failed, and I'll give another try with enabling logging and deleting the large cache file.

I can see the log in `Summary`.
```
Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-071420-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at System.Diagnostics.TraceSource.Flush()
   at GitHub.Runner.Common.TraceManager.Dispose(Boolean disposing)
   at GitHub.Runner.Common.TraceManager.Dispose()
   at GitHub.Runner.Common.HostContext.Dispose(Boolean disposing)
   at GitHub.Runner.Common.HostContext.Dispose()
   at GitHub.Runner.Worker.Program.Main(String[] args)
System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-071420-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
System.IO.IOException: No space left on device : '/home/runner/runners/2.299.1/_diag/Worker_20221214-071420-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Common.Tracing.Error(Exception exception)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
```

**Comment by dust1** [comment]
some case was fail, may I fix it latter

**Comment by ShiKaiWi** [comment]
@dust1 https://github.com/CeresDB/sqlness/pull/2 has fixed one issue which may lead to weird output when running sqlness tests. (I guess you may encounter the same problem when running tests locally)

**Comment by dust1** [comment]
@ShiKaiWi It should be all right now. I previously used `git merge main` to merge the changes of the main branch into my current branch. Is it possible that I missed something and the case modification of issue397 branch is not merged. small details slip up😂

**Comment by jiacai2050** [comment]
Never mind, general implementation seems good, I will push few small changes to your branch, then we can merge this.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**